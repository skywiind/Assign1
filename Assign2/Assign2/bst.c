/*****************************************************************************************
*                                           *                                            *
*             CS201-001 Assign 2            * This file defines the implementations of   *
*             Wyatt Fairbanks               * the BST class methods.                     *
*             Dr. John Lusth                *                                            *
*             bst.c                         *                                            *
*             10/22/2018                    *                                            *
*                                           *                                            *
*****************************************************************************************/
#include <stdio.h>
#include <stdlib.h>
#include <assert.h>
#include "bst.h"
#include "queue.h"

struct bst {
	void (*swap)    (TNODE *v1, TNODE *v2);
	int  (*compare) (void *v1, void *v2);
	void (*display) (void *value, FILE *fp);
	void (*free)    (void *value);
	TNODE *root;
	int    size;
	int    debug;
};

//UTILITY FUNCTIONS ////////////////////////////////////////////////////////////////

void swapNodes(TNODE *val1, TNODE *val2) {
	void *temp = getTNODEvalue(val1);
	setTNODEvalue(val1, getTNODEvalue(val2));
	setTNODEvalue(val2, temp);
	return;
}

int isLeaf(TNODE *n) {
	if (!n) {
		return 0;
	}
	if (getTNODEleft(n) == 0 && getTNODEright(n) == 0) {
		return 1;
	}
	return 0;
}

int min(int val1, int val2) {
	if (val1 > val2) {
		return val2;
	}
	if (val2 > val1) {
		return val1;
	}
	return -1;
}

int minDepth(TNODE *node) {
	if (node == 0) {
		return -1;
	}
	if (getTNODEleft(node) && getTNODEright(node)) {
		return min(minDepth(getTNODEleft(node)), minDepth(getTNODEright(node))) + 1;
	}
	return 0;
}

int max(int val1, int val2) {
	if (val1 > val2) {
		return val1;
	}
	if (val2 > val1) {
		return val2;
	}
	return -1;
}

int maxDepth(TNODE *node) {
	if (node == 0) {
		return -1;
	}
	if (isLeaf(node)) {
		return 0;
	}
	if (getTNODEleft(node) && !getTNODEright(node)) {
		return maxDepth(getTNODEleft(node)) + 1;
	}
	if (!getTNODEleft(node) && getTNODEright(node)) {
		return maxDepth(getTNODEright(node)) + 1;
	}
	if (getTNODEleft(node) && getTNODEright(node)) {
		return max(maxDepth(getTNODEleft(node)), maxDepth(getTNODEright(node))) + 1;
	}
	return 0;
}

int isLeftChild(TNODE *n) {
	if (!n) {
		return 0;
	}
	if (getTNODEleft(getTNODEparent(n)) == n) {
		return 1;
	}
	return 0;
}

int isRightChild(TNODE *n) {
	if (!n) {
		return 0;
	}
	if (getTNODEright(getTNODEparent(n)) == n) {
		return 1;
	}
	return 0;
}

void levelOrderDisplay(TNODE *n, FILE *fp) {
	QUEUE *queue = newQUEUE();
	TNODE *temp = n;
	int i = 0;
	int l = 0;
	int c = 1;
	while (l != maxDepth(n) + 2) {
		if (i == 0) {
			fprintf(fp, "%d : ", l);
			l++;
		}
		else if (i == c) {
			fprintf(fp, "\n%d : ", l);
			c *= 2;
			l++;
			i = 0;
		}
		if (temp) {
			if (isLeaf(temp) == 1) {
				fprintf(fp, "=");
			}
			displayTNODE(temp, fp);
			fprintf(fp, "(");
			displayTNODE(getTNODEparent(temp), fp);
			fprintf(fp, ")");
			if (i == 0 && l == 1) {
				fprintf(fp, "X ");
			}
			else if (isLeftChild(temp) == 1) {
				fprintf(fp, "L ");
			}
			else if (isRightChild(temp) == 1) {
				fprintf(fp, "R ");
			}
			enqueue(queue, getTNODEleft(temp));
			enqueue(queue, getTNODEright(temp));
		}
		else {
			enqueue(queue, NULL);
			enqueue(queue, NULL);
		}
		temp = dequeue(queue);
		i++;
	}
	freeQUEUE(queue);
	return;
}

void inOrderDisplay(BST *t, TNODE *n, FILE *fp) {
	if (getTNODEleft(n)) {
		fprintf(fp, "[");
		inOrderDisplay(t, getTNODEleft(n), fp);
		fprintf(fp, "] ");	
	}
	if (n) {
		setTNODEdisplay(n, t->display);
		displayTNODE(n, fp);
	}
	if (getTNODEright(n)) {
		fprintf(fp, " [");
		inOrderDisplay(t, getTNODEright(n), fp);
		fprintf(fp, "]");
	}
}

void preOrderDisplay(BST *t, TNODE *n, FILE *fp) {
	if (n) {
		setTNODEdisplay(n, t->display);
		displayTNODE(n, fp);
	}
	if (getTNODEleft(n)) {
		fprintf(fp, " [");
		preOrderDisplay(t, getTNODEleft(n), fp);
		fprintf(fp, "]");
	}
	if (getTNODEright(n)) {
		fprintf(fp, " [");
		preOrderDisplay(t, getTNODEright(n), fp);
		fprintf(fp, "]");
	}
}

void postOrderDisplay(BST *t, TNODE *n, FILE *fp) {
	if (getTNODEleft(n)) {
		fprintf(fp, "[");
		postOrderDisplay(t, getTNODEleft(n), fp);
		fprintf(fp, "] ");
	}
	if (getTNODEright(n)) {
		fprintf(fp, "[");
		postOrderDisplay(t, getTNODEright(n), fp);
		fprintf(fp, "] ");
	}
	if (n) {
		setTNODEdisplay(n, t->display);
		displayTNODE(n, fp);
	}
}

void postOrderFree(BST *t, TNODE *n) {
	if (getTNODEleft(n)) {
		postOrderFree(t, getTNODEleft(n));
	}
	if (getTNODEright(n)) {
		postOrderFree(t, getTNODEright(n));
	}
	if (n) {
		setTNODEfree(n, t->free);
		freeTNODE(n);
	}
}

//END UTILITY FUNCS ////////////////////////////////////////////////////////////////

BST *newBST(int(*c) (void *, void*)) {
	BST *tree     = malloc(sizeof(BST));
	tree->root    = 0;
	tree->size    = 0;
	tree->debug   = 0;
	tree->swap    = swapNodes;
	tree->compare = c;
	tree->display = 0;
	tree->free    = 0;
	return tree;
}

void setBSTdisplay(BST *t, void (*d)(void *, FILE *)) {
	t->display = d;
	return;
}

void setBSTswapper(BST *t, void (*s)(TNODE *, TNODE *)) {
	t->swap = s;
	return;
}

void setBSTfree(BST *t, void(*f)(void *)) {
	t->free = f;
	return;
}

TNODE *getBSTroot(BST *t) {
	return t->root;
}

void setBSTroot(BST *t, TNODE *replacement) {
	t->root = replacement;
	return;
}

void setBSTsize(BST *t, int s) {
	t->size = s;
	return;
}

TNODE *insertBST(BST *t, void *value) {
	TNODE *node    = t->root;
	TNODE *newNode = newTNODE(value, 0, 0, 0);
	if (t->display) {
		setTNODEdisplay(newNode, t->display);
	}
	if (node == 0) {
		setTNODEparent(newNode, newNode);
		t->root = newNode;
		setBSTsize(t, 1);
		return newNode;
	}
	while (node != 0) {
		int c = t->compare(getTNODEvalue(node), value);
		if (c > 0 && getTNODEleft(node) == 0) {
			setTNODEleft(node, newNode);
			setTNODEparent(newNode, node);
			setBSTsize(t, t->size + 1);
			return newNode;
		}
		if (c < 0 && getTNODEright(node) == 0) {
			setTNODEright(node, newNode);
			setTNODEparent(newNode, node);
			setBSTsize(t, t->size + 1);
			return newNode;
		}
		if (c > 0) {
			node = getTNODEleft(node);
		}
		if (c < 0) {
			node = getTNODEright(node);
		}
		if (c == 0) {
			return NULL;
		}
	}
	return NULL;
}

void *findBST(BST *t, void *value) {
	TNODE *node = t->root;
	if (node == 0) {
		return NULL;
	}
	while (node != 0) {
		int c = t->compare(getTNODEvalue(node), value);
		if (c == 0) {
			return getTNODEvalue(node);
		}
		else {
			if (c < 0) {
				node = getTNODEleft(node);
			}
			if (c > 0) {
				node = getTNODEright(node);
			}
		}
	}
	return NULL;
}

TNODE *locateBST(BST *t, void *value) {
	TNODE *node = t->root;
	if (node == 0) {
		return NULL;
	}
	while (node != 0) {
		int c = t->compare(getTNODEvalue(node), value);
		if (c == 0) {
			return node;
		}
		else {
			if (c < 0) {
				node = getTNODEleft(node);
			}
			if (c > 0) {
				node = getTNODEright(node);
			}
		}
	}
	return NULL;
}

int deleteBST(BST *t, void *key) {
	TNODE *target = locateBST(t, key);
	if (target != NULL) {
		return -1;
	}
	target = swapToLeafBST(t, target);
	pruneLeafBST(t, target);
	if (t->free) {
		t->free(getTNODEvalue(target));
	}
	free(target);
	setBSTsize(t, t->size - 1);
	if (t->size == 0) {
		t->root = NULL;
	}
	return 0;
}

TNODE *swapToLeafBST(BST *t, TNODE *node) {
	if (isLeaf(node) == 1) {
		return node;
	}
	if (getTNODEleft(node)) {
		t->swap(node, getTNODEleft(node));
		node = getTNODEleft(node);
		while (getTNODEright(node)) {
			t->swap(node, getTNODEright(node));
			node = getTNODEright(node);
		}
		return swapToLeafBST(t, node);
	}
	else if (getTNODEright(node)) {
		t->swap(node, getTNODEright(node));
		node = getTNODEright(node);
		while (getTNODEleft(node)) {
			t->swap(node, getTNODEleft(node));
			node = getTNODEleft(node);
		}
		return swapToLeafBST(t, node);
	}
	return node;
}

void pruneLeafBST(BST *t, TNODE *leaf) {
	TNODE *parent = getTNODEparent(leaf);
	if (getBSTroot(t) == leaf) {
		return;
	}	
	if (leaf == getTNODEleft(parent)) {
		setTNODEleft(parent, 0);
		return;
	}
	if (leaf == getTNODEright(parent)) {
		setTNODEright(parent, 0);
		return;
	}
}

int sizeBST(BST *t) {
	return t->size;
}

void statisticsBST(BST *t, FILE *fp) {
	int size = sizeBST(t);
	if (size == 0) {
		fprintf(fp, "Nodes: 0\nMinimum depth: -1\nMaximum depth: -1\n");
		return;
	}
	fprintf(fp, "Nodes: %d\nMinimum depth: %d\nMaximum depth: %d\n", size, minDepth(t->root), maxDepth(t->root));
	return;
}

void displayBST(BST *t, FILE *fp) {
	if (getBSTroot(t) == 0) {
		fprintf(fp, "0\n");
		return;
	}

	if (t->debug > 0 && t->root == NULL) {
		fprintf(fp, "[]\n");
		return;
	}
	else if (t->debug == 1) {
		fprintf(fp, "[");
		inOrderDisplay(t, t->root, fp);
		fprintf(fp, "]\n");
	}
	else if (t->debug == 2) {
		fprintf(fp, "[");
		preOrderDisplay(t, t->root, fp);
		fprintf(fp, "]\n");
	}
	else if (t->debug == 3) {
		fprintf(fp, "[");
		postOrderDisplay(t, t->root, fp);
		fprintf(fp, "]\n");
	}
	else {
		levelOrderDisplay(getBSTroot(t), fp);
	}
	return;
}

int debugBST(BST *t, int level) {
	int temp = t->debug;
	t->debug = level;
	return temp;
}

void freeBST(BST *t) {
	postOrderFree(t, t->root);
	free(t);
}