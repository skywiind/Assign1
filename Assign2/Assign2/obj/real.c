#include <stdio.h>
#include <stdlib.h>
#include <limits.h>
#include <assert.h>
#include "real.h"

struct REAL{
    double value;
};

REAL* newREAL(double x){
    REAL *p = malloc(sizeof(REAL));
    assert(p != 0);
    p->value = x;
    return p;
}

double getREAL(REAL *v){
    return v->value;
}

double setREAL(REAL *v,double x){
    double old = v->value;
    v->value = x;
    return old;
}

void displayREAL(void *v,FILE *fp){
    fprintf(fp,"%lf",getREAL((REAL*) v));
	return;
}

double compareREAL(void *v,void *w){
    return getREAL(v) - getREAL(w);
}

void freeREAL(void *v){
    free((REAL *) v);
	return;
}
