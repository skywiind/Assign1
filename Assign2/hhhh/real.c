#include <stdio.h>
#include <stdlib.h>
#include <limits.h>
#include <assert.h>
#include "real.h"

struct REAL{
    double value;
};

REAL* newREAL(double x){
    REAL *p = malloc(sizeof(REAL));
    assert(p != 0);
    p->value = x;
    return p;
}

double getREAL(REAL *v){
    return v->value;
}

double setREAL(REAL *v,double x){
    double old = v->value;
    v->value = x;
    return old;
}

void displayREAL(void *v,FILE *fp){
    fprintf(fp,"%lf",getREAL((REAL*) v));
	return;
}

int compareREAL(void *v,void *w){
    double r = getREAL(v) - getREAL(w);
    if (r < 0.0) {
	    return -1;
    } else if (r > 0.0) {
	    return 1;
    } else {
	    return 0;
    }
}

void freeREAL(void *v){
    free((REAL *) v);
	return;
}
