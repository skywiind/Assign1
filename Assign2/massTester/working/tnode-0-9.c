#include <stdio.h>
#include <stdlib.h>
#include "bst.h"
#include "integer.h"
#include "real.h"
#include "string.h"
#include "queue.h"

void srandom(unsigned int);
long int random(void);

int
main(void)
    {
    srandom(21);
    //simple INTEGER test of BST
    int i;
    BST *p = newBST(compareINTEGER);
    setBSTdisplay(p,displayINTEGER);
    setBSTfree(p,freeINTEGER);
    for (i = 0; i < 10000; ++i)
        {
        int j = random() % 10000;
        INTEGER *a = newINTEGER(j);
        if (findBST(p,a) == 0)
            insertBST(p,a);
        else
            freeINTEGER(a);
        }
    if (sizeBST(p) < 200)
        {
        debugBST(p,1);
        printf("in-order: ");
        displayBST(p,stdout);
        printf("\n");
        debugBST(p,0);
        printf("level:\n");
        displayBST(p,stdout);
        }
    printf("size is %d\n",sizeBST(p));
    while (sizeBST(p) > 200)
        {
        for (i = 0; i < 10000; ++i)
            {
            int j = random() % 10000;
            INTEGER *a = newINTEGER(j);
            INTEGER *b;
            if ((b = findBST(p,a)) != 0)
                {
                deleteBST(p,a);
                freeINTEGER(b);
                }
            freeINTEGER(a);
            }
        }
    if (sizeBST(p) < 200)
        {
        debugBST(p,1);
        printf("in-order: ");
        displayBST(p,stdout);
        printf("\n");
        debugBST(p,0);
        printf("level:\n");
        displayBST(p,stdout);
        }
    printf("size is %d\n",sizeBST(p));
    statisticsBST(p,stdout);
    freeBST(p);
    return 0;
    }
