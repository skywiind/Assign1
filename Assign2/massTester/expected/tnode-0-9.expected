size is 6374
in-order: [[[[[91 [[[108] 122] 184 [285]]] 386] 388] 622 [908 [934]]] 990 [[[[[1386 [[[[1583 [1665]] 1717] 1812] 1913]] 1927] 1968] 2003 [[[[2330] 2363] 2660] 2789 [[2950] 3002]]] 3031 [[[[3122] 3258] 3306 [[[[[3370] 3479] 3504 [[3730] 3747]] 3864 [[[4253 [4323 [4351]]] 4359 [[[4577] 4635] 4653 [4721]]] 4766 [4960]]] 5001 [[5149 [5294]] 5316 [[5389] 5533 [5656 [5715]]]]]] 5804 [[[[[[[5950] 5957] 5958] 6056] 6057] 6249 [[[[[[6350] 6361] 6475] 6771 [6909]] 6912 [[[[[6975] 7021] 7047 [7160 [7217]]] 7223] 7270]] 7396 [[7570] 7614 [7906 [7947]]]]] 8157 [[[[8166 [8360 [[8392 [8466 [8481]]] 8656 [8755]]]] 8801 [8864]] 8932 [[8986] 9178 [9278]]] 9281 [[[[9421 [9439]] 9475 [9520]] 9542 [[9583 [[9724] 9748]] 9767 [[9891] 9894]]] 9936]]]]]]
level:
0: 990(990)X
1: 622(990)L 3031(990)R
2: 388(622)L 908(622)R 2003(3031)L 5804(3031)R
3: 386(388)L =934(908)R 1968(2003)L 2789(2003)R 3306(5804)L 8157(5804)R
4: 91(386)L 1927(1968)L 2660(2789)L 3002(2789)R 3258(3306)L 5001(3306)R 6249(8157)L 9281(8157)R
5: 184(91)R 1386(1927)L 2363(2660)L =2950(3002)L =3122(3258)L 3864(5001)L 5316(5001)R 6057(6249)L 7396(6249)R 8932(9281)L 9936(9281)R
6: 122(184)L =285(184)R 1913(1386)R =2330(2363)L 3504(3864)L 4766(3864)R 5149(5316)L 5533(5316)R 6056(6057)L 6912(7396)L 7614(7396)R 8801(8932)L 9178(8932)R 9542(9936)L
7: =108(122)L 1812(1913)L 3479(3504)L 3747(3504)R 4359(4766)L =4960(4766)R =5294(5149)R =5389(5533)L 5656(5533)R 5958(6056)L 6771(6912)L 7270(6912)R =7570(7614)L 7906(7614)R 8166(8801)L =8864(8801)R =8986(9178)L =9278(9178)R 9475(9542)L 9767(9542)R
8: 1717(1812)L =3370(3479)L =3730(3747)L 4253(4359)L 4653(4359)R =5715(5656)R 5957(5958)L 6475(6771)L =6909(6771)R 7223(7270)L =7947(7906)R 8360(8166)R 9421(9475)L =9520(9475)R 9583(9767)L 9894(9767)R
9: 1583(1717)L 4323(4253)R 4635(4653)L =4721(4653)R =5950(5957)L 6361(6475)L 7047(7223)L 8656(8360)R =9439(9421)R 9748(9583)R =9891(9894)L
10: =1665(1583)R =4351(4323)R =4577(4635)L =6350(6361)L 7021(7047)L 7160(7047)R 8392(8656)L =8755(8656)R =9724(9748)L
11: =6975(7021)L =7217(7160)R 8466(8392)R
12: =8481(8466)R
size is 106
Nodes: 106
Minimum depth: 2
Maximum depth: 12
